### PR TITLE
Add e4s tags 

### PR DESCRIPTION
**Pull request by shahzebsiddiqui**
@sameershende @eugeneswalker 
This PR will add the e4s tags, the `ecp` tag is already registered with some packages so we just adding it assuming these packages are part of E4S and ECP. I used this [spack.yaml](https://github.com/E4S-Project/e4s/blob/master/spack.yaml) as a reference for identifying packages that are part of E4S. If this makes sense by all means we can start tagging the packages with `e4s`. This can be useful with `spack find -t e4s` to filter packages from non-e4s.

```
bash-3.2$ spack list -t e4s
==> 8 packages.
adios  adios2  aml  amrex  arborx  argobots  ascent  axom
```

```
bash-3.2$ spack list -t ecp
==> 29 packages.
adios   aml       amrex   argobots  axl   chombo            er    kvtree  latte      nek5000  nektools  parsplice  rankstr  shuffile  veloc
adios2  amr-wind  arborx  ascent    axom  ecp-data-vis-sdk  filo  lammps  nalu-wind  nekrs    nwchem    qmcpack    redset   spath

```

### CODE REVIEW

**Comment by eugeneswalker** [comment]
Seems like a good idea!

**Comment by sethrj** [comment]
@shahzebsiddiqui Should this be marked as a draft until all the packages are submitted? It sounds like a good idea to me but this PR is incomplete.

**Comment by shahzebsiddiqui** [comment]
> @shahzebsiddiqui Should this be marked as a draft until all the packages are submitted? It sounds like a good idea to me but this PR is incomplete.

Well i didn't intend on updating all the E4S packages with the tags in this PR so i think this PR could be considered complete unless there is an objection. I would think that it should be maintainers adding appropriate tags, but the use of `e4s` tag helps us facility staff that deploy E4S pinpoint E4S from non-e4s packages. I think @eugeneswalker can help track state of E4S packages and update these tags as needed.

**Comment by shahzebsiddiqui** [comment]
@sethrj if you would like me to do all the e4s packages I can do that it would update alot more files. I think @eugenewalker knows when a package becomes E4S.

**Comment by sethrj** [comment]
I think it would be good to get at least all the e4s codes tagged. Better to update them all at once IMHO.

**Comment by shahzebsiddiqui** [comment]
@sethrj sorry i just got around to doing this task. I based these changes off E4S 21.05 spack.yaml https://github.com/E4S-Project/e4s/blob/master/spack.yaml provided by @eugeneswalker . There were some packages in this list that didn't make sense why they were in e4s. Such as `gmp`? I think `py-*` extensions like `py-jupyterhub` doesn't make sense either but i did it anyways.  

Anyhow i did most of them please review them. I think i will leave this up to maintainers or E4S team to add tags when packages become E4S products.

**Comment by alalazo** [comment]
@shahzebsiddiqui CI is not passing, due to style and supported Python versions (here it seems you use constructs somewhere that are for Python 3.6 or later)

File: var/spack/repos/builtin/packages/gasnet/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/gasnet/package.py]: Tricky: `:` should be ` =`

**Review by sethrj — CHANGES_REQUESTED** [review]
Those `vermin` results sure don't make it easy to find the problem file. I've tagged the error for you.

The only quibble I have is that while all of these are `e4s` specced packages, not all of them are exactly ECP software: for example, ECP doesn't fund qt or gmp development. I think the `ecp` tag would be more useful if it *only* applies to packages that are directly funded by the project. Maybe @eugeneswalker could weigh in there.

File: var/spack/repos/builtin/packages/gasnet/package.py
Comment (shahzebsiddiqui) [reply var/spack/repos/builtin/packages/gasnet/package.py]: thanks ill fix that.

Comment (shahzebsiddiqui) [reply var/spack/repos/builtin/packages/gasnet/package.py]: this is fixed now in https://github.com/spack/spack/pull/23382/commits/a482c47ca5903eab15f28efb83b032ad57cb8d6c

**Comment by shahzebsiddiqui** [comment]
> @shahzebsiddiqui CI is not passing, due to style and supported Python versions (here it seems you use constructs somewhere that are for Python 3.6 or later)

@alalazo flake8 should be addressed now i tested this locally will check the CI job.

File: var/spack/repos/builtin/packages/argobots/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/argobots/package.py]: 🤔 Is this actually an ECP project?

Comment (sethrj) [reply var/spack/repos/builtin/packages/argobots/package.py]: Nor is qt, but they seem to be explicit packages in e4s... I agree the ecp tag should be dropped if the code development isn't at least partly funded by ecp.

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/argobots/package.py]: It appears in slides BUT not listed as an ST product.

File: var/spack/repos/builtin/packages/axom/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/axom/package.py]: Note this appears in slides but is not currently listed as an ST product.

Comment (eugeneswalker) [reply var/spack/repos/builtin/packages/axom/package.py]: Definitely part of E4S. Not sure about ECP

Comment (shahzebsiddiqui) [reply var/spack/repos/builtin/packages/axom/package.py]: okay i removed `ecp` tag from axom package

File: var/spack/repos/builtin/packages/argobots/package.py
Comment (shahzebsiddiqui) [reply var/spack/repos/builtin/packages/argobots/package.py]: okay `ecp` tag removed from argobots

**Comment by shahzebsiddiqui** [comment]
Please feel free to comment on all the packages that are part of `e4s` and or `ecp`. For `e4s` it was pretty straightforward. I just referenced this https://github.com/E4S-Project/e4s project and see its `spack.yaml` though i am not sure if some packages were missed from older versions of E4S.

The point of this PR is to allow us to run `spack list -t e4s` or even `spack find -t e4s`  in a deployment to see the root specs it helps to clean up the noise with all the dependent packages shown in `spack find`. This could potentially be useful for @wspear who is trying to make sense of the packages output I provide him such as this one for E4S 21.02 https://github.com/spack/spack-configs/blob/main/NERSC/cori/e4s-21.02/e4s-21.02.txt and process this into some dashboard. We could potentially run the same command with `-t <e4s>` to filter by e4s tags if we want to be explicit. Though `spack find` is still good enough as it reports everything.

**Comment by sethrj** [comment]
Unfortunately we can't merge this now because of https://github.com/spack/spack/pull/23947 😭 it touches too many packages, and it's not fair to make you go back through and edit them all by hand.  It looks like https://github.com/spack/spack/pull/24071 provides a potential pathway to an automated solution (@cosmicexplorer).

I *do* think you should remove the `ecp` tag from all the packages on this PR that you don't explicitly know are funded by ECP; since you're picking up at least a couple that definitely aren't ECP, making the tag pretty useless.

**Comment by shahzebsiddiqui** [comment]
@sethrj Do you want me to remove the `ecp` tag from existing packages before this PR i think there was about a dozen or so packages that have this. Sure i can remove the `ecp` tags from the ones I specified so far. Ill get back to you by tomorrow

**Comment by sethrj** [comment]
> @sethrj Do you want me to remove the `ecp` tag from existing packages before this PR i think there was about a dozen or so packages that have this. Sure i can remove the `ecp` tags from the ones I specified so far.

I would say to assume that people who individually tagged ecp did so for a reason besides "bundled with e4s" so just remove the ecp tags from the ones you're adding to this pr.

**Comment by shahzebsiddiqui** [comment]
I have removed the `ecp` tag from products that i know dont belong to ECP. I used one of our dashboards in our ECP confluence https://confluence.exascaleproject.org/display/HISD/SI+Build+and+Test+Dashboard to keep track of this @tldahlgren is one of the owners for this dashboard along with Ryan Adamson. I reviewed all of the specs i updated and removed package that were not in the list. This will address several packages that are part of ECP project but didnt have a `ecp` tag. This way we get both of these tags registered in one shot. I am sure i missed some packages it is up to maintainers to add these tags which should be relatively easy. Perhaps one of the E4S teams @eugeneswalker @wspear can update these tags when products are part of E4S with `e4s` tag. 

Not sure if that maintainer bot @adamjstewart created can help ping all the maintainers now to confirm. I think most products covered in this PR are part of ECP and E4S.

**Comment by adamjstewart** [comment]
The way the maintainer bot is designed, it won't ping people for review unless it's a new PR being created. This prevents the bot from pinging the exact same people every time a new commit is added. @tgamblin is working on a plugin that will allow you to manually run the bot by commenting with @spackbot, but I don't think it's ready yet.

**Comment by sethrj** [comment]
@shahzebsiddiqui Thankfully @tgamblin just merged a fix for *all* the import orderings, which means you should be able to merge/rebase and get a passing CI. 😅

**Comment by adamjstewart** [comment]
Not yet, #24695 still hasn't been merged.

**Comment by sethrj** [comment]
> Not yet, #24695 still hasn't been merged.

Whoops, I must've mistaken the "tgamblin push" on my homepage: I assumed it was to `develop` but it probably shows updates for feature branches too 🙄

**Comment by shahzebsiddiqui** [comment]
I rebased my branch and fixed all the merge conflicts. Looks like the import errors in style are still going to be there https://github.com/spack/spack/pull/23382/checks?check_run_id=3010350761 until we get this merged. Let me know once this is merged and i can rebase again

**Comment by sethrj** [comment]
Ugh. Sorry about all the hassle. 😞

**Comment by shahzebsiddiqui** [comment]
Okay i wanted to point out we now have some direction onto E4S products. We will start to document this in the e4s project at https://github.com/E4S-Project/e4s which will be the source of truth. There is an open PR: https://github.com/E4S-Project/e4s/pull/19 that i am working on which requires review from E4S team. Once this is merged, I can update this MR with the list of `e4s` tags. As per `ecp` tags ill need to check on that list and get back to you with an update

**Comment by shahzebsiddiqui** [comment]
@tldahlgren 
cc: @sameershende 

I have finally got all the packages updated. I have removed `ecp` tag from all packages that I added. I have updated all packages with `e4s` tag according to https://github.com/E4S-Project/e4s/blob/master/E4S_Products.md which is the official list for e4s products. This list may change over time in that case E4S team can go in and update the spack package with the `e4s` tag. Currently we have 102 packages with e4s tag. There are total of 88 products in the E4S product table however some products have more than one spack package such as `darshan`, `flux`, `llvm`, `superlu`, etc which is why we have this difference.

Please review this and merge this when its ready.

```
bash-3.2$ spack list -t e4s
==> 102 packages.
adios     bolt             dyninst      globalarrays    libnrm         mpich             papyrus          py-cinemasci    scr           tasmanian  warpx
adios2    cabana           faodel       gotcha          libquo         mpifileutils      parallel-netcdf  py-jupyterhub   slate         tau        zfp
aml       caliper          flecsi       hdf5            llvm           netlib-scalapack  paraview         py-libensemble  slepc         trilinos
amrex     chai             flit         heffte          llvm-doe       ninja             parsec           py-petsc4py     stc           umap
aomp      charliecloud     flux-core    hpctoolkit      loki           nrm               pdt              py-warpx        strumpack     umpire
arborx    conduit          flux-sched   hpx             magma          nvhpc             petsc            qt              sundials      unifyfs
archer    darshan-runtime  fortrilinos  hypre           mercury        omega-h           phist            qthreads        superlu       upcxx
argobots  darshan-util     gasnet       kokkos          metall         openmpi           plasma           qwt             superlu-dist  variorum
ascent    datatransferkit  geopm        kokkos-kernels  mfem           openpmd-api       precice          raja            swig          veloc
axom      dealii           ginkgo       legion          mpark-variant  papi              pumi             rempi           sz            vtk-m
```

File: var/spack/repos/builtin/packages/gasnet/package.py
Comment (bonachea) [inline var/spack/repos/builtin/packages/gasnet/package.py]: Why was this whitespace / paragraph break removed from our package description?

This is here to emphasize the notice below.

**Review by bonachea — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/mfem/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/mfem/package.py]:
```suggestion
    tags = ['fem', 'finite-elements', 'high-order', 'amr', 'hpc', 'radiuss', 'e4s']
```

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/mfem/package.py]: Don't drop the `radiuss` tag.

File: var/spack/repos/builtin/packages/mpich/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/mpich/package.py]: Style issue.

```suggestion

```

File: var/spack/repos/builtin/packages/sz/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/sz/package.py]: This appears to be out of scope for this PR.  Was its addition here intentional?

**Review by tldahlgren — CHANGES_REQUESTED** [review]

**Comment by sethrj** [comment]
@shahzebsiddiqui Looks like some rebases might have gone horribly awry.

File: var/spack/repos/builtin/packages/mfem/package.py
Comment (shahzebsiddiqui) [reply var/spack/repos/builtin/packages/mfem/package.py]: that was my mistake. Will add this in

File: var/spack/repos/builtin/packages/mpich/package.py
Comment (shahzebsiddiqui) [reply var/spack/repos/builtin/packages/mpich/package.py]: good catch. I will fix this

File: var/spack/repos/builtin/packages/sz/package.py
Comment (shahzebsiddiqui) [reply var/spack/repos/builtin/packages/sz/package.py]: that was not intention. That was me rebasing my branch and resulted in merge conflict. anyhow i checked the sz package https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/sz/package.py and it doesn't have this section so i removed it

File: var/spack/repos/builtin/packages/mpich/package.py
Comment (shahzebsiddiqui) [reply var/spack/repos/builtin/packages/mpich/package.py]: i fixed all the style issues reported in CI check. There were a few more packages that had same issue.

File: var/spack/repos/builtin/packages/gasnet/package.py
Comment (shahzebsiddiqui) [reply var/spack/repos/builtin/packages/gasnet/package.py]: i added the new line in https://github.com/spack/spack/pull/23382/commits/43013f54f71db4fb2d517225cad0b421564841cc.

The newline doesn't show up in `spack info` so it doesn't make a difference apart from readability. 

```
bash-3.2$ spack info gasnet
Package:   gasnet

Description:
    GASNet is a language-independent, networking middleware layer that
    provides network-independent, high-performance communication primitives
    including Remote Memory Access (RMA) and Active Messages (AM). It has
    been used to implement parallel programming models and libraries such as
    UPC, UPC++, Co-Array Fortran, Legion, Chapel, and many others. The
    interface is primarily intended as a compilation target and for use by
    runtime library writers (as opposed to end users), and the primary goals
    are high performance, interface portability, and expressiveness.
    ***NOTICE***: The GASNet library built by this Spack package is ONLY
    intended for unit-testing purposes, and is generally UNSUITABLE FOR
    PRODUCTION USE. The RECOMMENDED way to build GASNet is as an embedded
    library as configured by the higher-level client runtime package (UPC++,
    Legion, etc), including system-specific configuration.
```

File: var/spack/repos/builtin/packages/mpich/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/mpich/package.py]: Whitespace error

**Review by sethrj — DISMISSED** [review]
Looks good but I think you have a whitespace error that will cause the style check to fail.

File: var/spack/repos/builtin/packages/gasnet/package.py
Comment (bonachea) [reply var/spack/repos/builtin/packages/gasnet/package.py]: > The newline doesn't show up in spack info so it doesn't make a difference apart from readability.

This is a _defect_ in the current implementation of `spack info`, reported as issue #23831 that will hopefully be corrected soon.

Thanks for restoring the formatting that will hopefully be rendered correctly in an upcoming release.

File: var/spack/repos/builtin/packages/mpich/package.py
Comment (shahzebsiddiqui) [reply var/spack/repos/builtin/packages/mpich/package.py]: all style checks have been addressed and I checked the latest CI check and i saw no errors.

**Review by sethrj — APPROVED** [review]
Looks good, the only changes are adding `e4s` tags. @tldahlgren this will need your re-approval before it's merged.

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**
